### PR TITLE
WIP commit with the current status of "Analysis with python" project

### PR DESCRIPTION
**Pull request by renaudin**
This is a commit with all the files copied from the repository
used during the project.

There are a jupyter notebook, a script (with almost the same features)
and some plots.

### CODE REVIEW

File: contrib-03-AnalysisWithPython/Readme.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+## Motivation
+
+In High Energy Physics, almost every data analysis framework involved the ROOT
+software. Built in the 90's, it has been very useful to perform all the tasks
+required by an analysis, from the handle of large datasets to the use of
+powerful statistical procedures with the help of librairies such as RooFit or
+TMVA.  
+However, it hasn't spread outside of the HEP community while other projects
+also dedicated to data analysis are now very mature with a much larger
+users community.
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md:12]: Add a comma “HEP community, while other”.

File: contrib-03-AnalysisWithPython/Readme.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+The Goal of this project was to build an analysis demonstrator capable of :
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md]: Lowercase “G”.

File: contrib-03-AnalysisWithPython/Readme.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      file and store the content in a pandas dataframe. Follow both links if
+      you want more information on how to load a ROOT file. To install it, do
+      `pip install root_pandas`
+    - In order to produce a jupyter notebook which is a way to keep both the
+      analysis code and its output in the same file, you can also do `pip install jupyter`. Then, to load a jupyter notebook, you can do
+      `jupyter notebook` and access the notebook in your browser.
+
+## Loading data
+
+Using root_pandas library, we can load a ROOT file in a few simple steps :
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md]: “Using _the_ root_pandas…”

File: contrib-03-AnalysisWithPython/Readme.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+- choose the file you want and store the path as string : `myfile = "path/to/file.root"`
+- then select the branches you want by creating a list with the names of the
+  variables you want. The use of of the wildcard **\*** is also supported. For
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md]: could use backticks here to prevent the need to escape everything, i.e. `*`.

File: contrib-03-AnalysisWithPython/Readme.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  instance you can select `mybranches = ["B0_M","*_PT"]` to load the B0 mass
+  and the transverse momentum of each particle of your entry.
+- As the file will be loaded in the memory, you should limit to a certain
+  number of entries and set, for instance, `n_events = 10000`.
+- Finally, select the name of your tuple as `tuple_name = 'someDir/myTree`
+
+Then, you will be allowed to store you tree as a pandas dataframe using :
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md]: “_your_ tree”

File: contrib-03-AnalysisWithPython/Readme.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+a way to both :
+- create a reduce dataset : 
+    ```python
+    cut_string = " B0_M >5250 & B0_P > 500 "
+    df_reduced = df.query(cut_string)
+    ```   
+- plot an histogram using a selection cut :
+    ```python
+        df.query(cut_string)['B0_M'].plot.hist(bins=custom_binning)
+    ```
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md:131]: We could add a caveat here: [`DataFrame.query`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.query.html) will return a copy of the data by default. You can instead generate a ‘view’ by indexing with a boolean mask:

```
mask = (df['B0_M'] > 5250) & (df['B0_P'] > 500)
df_reduced = df[mask]
```

The query string is easier to express, but the `query` call will take up more memory as you end up with two `DataFrame` objects lying around. Which to use depends on the use-case 😄

File: contrib-03-AnalysisWithPython/Readme.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                            n_estimators=800, learning_rate=0.5)
+    #Training
+    bdt.fit(X_train, y_train)
+    
+    #BDT output
+    sk_y_predicted = bdt.predict(X_test) 
+```
+## Adding variable to the tree and saving output
+
+Some analysis step could create a new variable, as a BDT output, worth to be
```
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md]: “_such_ as a BDT output”, “_which is_ to be added…”

**Comment by alexpearce** [comment]
Thank you! This is very neat indeed. I've added a few comments, mostly editorial.

**Comment by renaudin** [comment]
Thank for your comments, I took them into account and correct a few things. The note is almost ready and Mick and I are still working to give a well documented script.

**Comment by alexpearce** [comment]
Thanks for the corrections, LGTM 👍  Looking forward to seeing the script! 👀

File: contrib-03-AnalysisWithPython/Readme.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
+# Analysis with python
+
+## Motivation
+
+In High Energy Physics, almost every data analysis framework involved the ROOT
+software. Built in the 90's, it has been very useful to perform all the tasks
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md:6]: Add link to ROOT: https://root.cern.ch/

File: contrib-03-AnalysisWithPython/Readme.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+## Environment setup
+
+In order to install and load all the packages we need for this project, we rely
+on the [anaconda python distribution](https://www.continuum.io/downloads).
+
+We tested the following environment on lxplus :
+
+- Install miniconda (or anaconda) for python2.7, Linux 64 bit
+  [link](http://conda.pydata.org/miniconda.html)
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: Maybe give the used commands for the install?

File: contrib-03-AnalysisWithPython/Readme.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+Using the root_pandas library, we can load a ROOT file in a few simple steps :
+- choose the file you want and store the path as string : `myfile = "path/to/file.root"`
+- then select the branches you want by creating a list with the names of the
+  variables you want. The use of of the wildcard `*` is also supported. For
+  instance you can select `mybranches = ["B0_M","*_PT"]` to load the B0 mass
+  and the transverse momentum of each particle of your entry.
+- As the file will be loaded in the memory, you should limit to a certain
+  number of entries and set, for instance, `n_events = 10000`.
+- Finally, select the name of your tuple as `tuple_name = 'someDir/myTree`
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: Maybe the code corresponding to all these steps could be moved to its own code block after the enumeration with its own comments to help:

``` python
# Path of the ROOT file
myfile = "path/to/file.root"
# Branches to load, wildcards allowed
mybranches = ["B0_M", "*_PT"]
# Number of entries to load
n_events = 10000
# Name of the tree
tuple_name = "someDir/myTree"
```

File: contrib-03-AnalysisWithPython/Readme.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+Then, you will be allowed to store your tree as a pandas dataframe using :
+```python
+   import root_pandas
+   df = root_pandas.read_root(myfile, columns=mybranches, stop = n_events, key= tuple_name)
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: PEP8: `df = root_pandas.read_root(myfile, columns=mybranches, stop=n_events, key=tuple_name)`

File: contrib-03-AnalysisWithPython/Readme.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+```
+## Creating and plotting histograms
+
+To produce a simple plot, let's say, the mass of the B0, you can simply do :
+```python
+   import numpy as np
+   custom_binning = np.linspace(5000,5500,100) #min, max and number of bins 
+   mass_B0 = df['B0_M'].plot.hist(bins=custom_binning, label = 'mass of the
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md:93]: I think it's better if you use `B0_MM`, that's the recommended mass to use.

File: contrib-03-AnalysisWithPython/Readme.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+- Finally, select the name of your tuple as `tuple_name = 'someDir/myTree`
+
+Then, you will be allowed to store your tree as a pandas dataframe using :
+```python
+   import root_pandas
+   df = root_pandas.read_root(myfile, columns=mybranches, stop = n_events, key= tuple_name)
+```
+## Creating and plotting histograms
+
+To produce a simple plot, let's say, the mass of the B0, you can simply do :
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: Use LaTeX for the B0 -> $B^{0}$

File: contrib-03-AnalysisWithPython/Readme.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+   mass_B0 = df['B0_M'].plot.hist(bins=custom_binning, label = 'mass of the
+   B^{0}_{M} [MeV/c$^{2}$]'} #label of the plot can be written in LaTeX
+```
+
+Then, you just need to add `plt.show()` in your code to plot the histogram.
+A more detailed code is written in both `compare_mc_data.py` and
+`AnalysisWithPython.ipynb`.  
+
+
+## Cutting
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: Cutting -> Selecting events

In principle, "cut" is jargon so we should try to avoid it as much as possible

File: contrib-03-AnalysisWithPython/Readme.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    ```python
+    cut_string = " B0_M >5250 & B0_P > 500 "
+    df_reduced = df.query(cut_string)
+    ```   
+- select events without creating a new dataset :
+    ```python
+    mask = (df['B0_M'] > 5250) & (df['B0_P'] > 500)
+    df_reduced = df[mask]
+    ```
+Which method to choose will depend on the use-case. And then, the plotting part
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md:127]: Can you expand in the pros and cons of each method?

File: contrib-03-AnalysisWithPython/Readme.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+  local python distribution as clean as possible. It can be achieve trough
+  : `conda create -n NameOfYourVirtualEnv python=2.7 anaconda`.
+  To activate it, do `source activate NameOfYourVirtualEnv` and `source deactivate` to go back to normal. 
+- Install ROOT inside the conda environment : `conda install -c chrisburr root=6.06.02`
+- You will then need to activate the conda environment :
+    - `source activate root` to load the main conda environment
+    - `source activate NameOfYourVirtualEnv` if you made it yourself
+- A few packages are also needed and can be installed using pip, the python
+  package manager :
+    - Numpy (arrays and tools for scientific computing), Matplotlib (plotting
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: References (links) to all packages are needed

File: contrib-03-AnalysisWithPython/Readme.md [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    #BDT output
+    sk_y_predicted = bdt.predict(X_test) 
+```
+## Adding variable to the tree and saving output
+
+Some analysis step could create a new variable, such as a BDT output, which is to be
+added to a ROOT file. Two steps are needed to perform this :
+- The variable should be added to the dataframe using `df['BDT'] = sk_y_predicted`
+- The modified dataframe can be stored in a new ROOT file using
+  `df.root_pandas.to_root(df, 'output.root', key = 'mytree')`
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: Same comment as before: maybe add the code as a block at the end of the enumeration?

``` python
df['BDT'] = sk_y_predicted
df.root_pandas.to_root(df, 'output.root', key='mytree')
```

Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md]: Could you add a few lines on how to store the extra variable back to disk, either as a ROOT file or as a Pandas dataframe? That would close the loop.

File: contrib-03-AnalysisWithPython/Readme.md [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+
+## Conclusion
```
Comment (apuignav) [inline contrib-03-AnalysisWithPython/Readme.md:194]: Before the conclusion, I think it would be nice to have a "Going further" (short) section in which you outline how to continue with the analysis, maybe giving links to useful tools for python-based analysis.

**Comment by apuignav** [comment]
Great work! I added my comments inline.

As a final comment, maybe we could change the title to "Getting started with python-based analysis"? This is more adequate to the content and makes it maybe more useful for people reading the proceedings.

File: contrib-03-AnalysisWithPython/Readme.md (hunk 12 above)
Comment (alexpearce) [inline contrib-03-AnalysisWithPython/Readme.md:93]: No, it's almost always the `_M` branch one should use. There was a [thread on `lhcb-tesla` about it](https://groups.cern.ch/group/lhcb-tesla/Lists/Archive/Flat.aspx?RootFolder=%2Fgroup%2Flhcb-tesla%2FLists%2FArchive%2Fdifference%20between%20XX_MM%20and%20XX_M&FolderCTID=0x01200200D58CE2C7A01AE94382C3D6890D64B854) a while ago, where Vanya said:

> Honestly speaking I personally do not understand why people are using MM at all...  - it could be useful for unndertsanding of vertex constrains, but fro all physics purposed, clearly  only "M" should be used  (or , much better, mass  after full DTF-refit)

**Comment by alexpearce** [comment]
:+1: on the title change.

**Comment by renaudin** [comment]
I will look at the thread you mentionned, and I agree, the proposed title is much better.

**Comment by renaudin** [comment]
About the Going further section and adding links to good material, I was thinking about pointing to :
- Software Carpentry lessons
- Starterkit lessons ;)
- Tim's blog

Do you have other links you would like to see in this section ?

**Comment by apuignav** [comment]
I'm not sure these are the most useful links. For full python-based analysis, maybe
- Theano/tensorflow
- Matplotlib
- The Impactkit Papermake article (#1)

What else? @ibab, @alexpearce?

**Comment by alexpearce** [comment]
I would add:
- [scipy](https://www.scipy.org/) for general scientific computation and numerical routines (such as kernel density estimate, spline interpolation, minimisation, etc.)
- The [uncertainties](https://pythonhosted.org/uncertainties/) and/or [mcerp](https://github.com/tisimst/mcerp) packages for dealing with error propagation
- The [iminuit](https://github.com/iminuit/iminuit) and/or [probfit](https://github.com/iminuit/probfit) packages for fitting

I think Theano/Tensorflow might be a bit daunting to start with.

**Comment by apuignav** [comment]
Completely agree with the list. Maybe we could add [soerp](https://pypi.python.org/pypi/soerp) for error propagation too?

**Comment by alexpearce** [comment]
I haven't used soerp, but we could add it (I think mcerp trumps just about anything, but it might be good to show the choice…).

**Comment by apuignav** [comment]
Never used either, I'm an `uncertainties` guy (I even wrote some code for it), but now I will have to try 😄
